Source/UticaKit/Project.swift

```diff
@@ -1130,127 +1130,247 @@ public final class Project { // swiftlint:disable:this type_body_length
       }
   }
 
+  struct ConcurrentBuildState {
+    typealias Observer = (Signal<ResolvedDependency, CarthageError>.Observer, Lifetime)
+    var observer: Observer? {
+      didSet {
+        if observer != nil {
+          sendNextChunk()
+        }
+      }
+    }
+    private var builtDependencies: Set<Dependency>
+    private var dependencies: [BuildGraphNode]
+
+    init(dependencies: [BuildGraphNode], builtDependencies: Set<Dependency>) {
+      self.dependencies = dependencies
+      self.builtDependencies = builtDependencies
+      self.observer = nil
+    }
+
+    mutating func markAsBuilt(dependency: Dependency) {
+      builtDependencies.insert(dependency)
+      sendNextChunk()
+    }
+
+    private mutating func sendNextChunk() {
+      guard let (observer, lifetime) = observer else {
+        return
+      }
+
+      guard !lifetime.hasEnded else {
+        self.observer = nil
+        return
+      }
+
+      let pivot = dependencies.partition { $0.dependencies.isSubset(of: builtDependencies) }
+      dependencies[pivot...]
+        .forEach {
+          if !lifetime.hasEnded {
+            observer.send(value: $0.resolvedDependency)
+          }
+        }
+      dependencies.removeSubrange(pivot...)
+      if dependencies.isEmpty {
+        if !lifetime.hasEnded {
+          observer.sendCompleted()
+        }
+        self.observer = nil
+      }
+    }
+  }
+
+  private func concurrentBuildDependencies(
+    _ dependenciesToBuild: [BuildGraphNode],
+    prebuiltDependencies: Set<Dependency>,
+    options: BuildOptions,
+    sdkFilter: @escaping SDKFilterCallback = { sdks, _, _, _ in .success(sdks) }
+  ) -> BuildSchemeProducer {
+    let numberOfJobs = options.jobs > 0 ? UInt(options.jobs) : UInt(ProcessInfo.processInfo.processorCount)
```

Comment:
Let's test performance, but I think `ProcessInfo.processInfo.processorCount` returns number of logical CPUs and in case of HyperThreat technology it will return 8 instead of 4 for Intel CPUs.
We need to perform some tests with different values and find the best default value. It may be number of physical CPUs

Reply:
Yeah, running few tests is a good idea. If we do so – it makes most of sense for us to do it on CI

Trick is – i think optimal job count depends on environment; CI is most of chances will run at best with `processorCount` while on Intel we might want to have 0.75x`processorCount`, and on M1 – maybe even 1.5x`processorCount`

Reply:
also, we can kinda create a benchmarking table in README with Date, cpu type, jobs count, time and contribute there :)

Reply:
Already did that. It was the intention to utilize all logical cores, because it actually adds some additional performance on top of the physical cores.

Results:
```sh
All - carthage - 997.27s user 475.96s system 34% cpu 1:11:23.90 total
All - utica --jobs 4 -  1264.20s user 566.71s system 102% cpu 29:52.79 total
All - utica --jobs 8 -  1338.80s user 603.35s system 128% cpu 25:05.70 total
```

Although, the difference is not dramatic, it gives another 7% performance.

Reply:
almos 3x with 8 jobs!

Reply:
I'm running it and it is AMAZING!